include/fmt/core.h

```diff
@@ -1613,6 +1629,194 @@ inline void print(const S& format_str, Args&&... args) {
+  could be shorter than the format arguments object.
+
+  It can be implicitly converted into `~fmt::basic_format_args` for passing
+  into type-erased formatting functions such as `~fmt::vformat`.
+  \endrst
+ */
+template <typename Context>
+class dynamic_format_arg_store
+#if FMT_GCC_VERSION < 409
+    // Workaround a GCC template argument substitution bug.
+    : public basic_format_args<Context>
+#endif
+{
+ private:
+  using char_type = typename Context::char_type;
+  using string_type = std::basic_string<char_type>;
+  using value_type = basic_format_arg<Context>;
+
+  template <typename T>
+  using stored_type =
+      conditional_t<internal::is_string<T>::value, string_type, T>;
+
+  // Storage of basic_format_arg must be contiguous
+  // Required by basic_format_args::args_ which is just a pointer
+  std::vector<value_type> data_;
+
+  // Storage of arguments not fitting into basic_format_arg must grow
+  // without relocation because items in data_ refer to it.
+
+  internal::dyn_arg_storage storage_;
+
+  friend class basic_format_args<Context>;
+
+  unsigned long long get_types() const {
+    return internal::is_unpacked_bit | data_.size();
+  }
+
+  template <typename T> const T& stored_value(const T& arg, std::false_type) {
+    return arg;
+  }
+
+  template <typename T>
+  const T& stored_value(const std::reference_wrapper<T>& arg, std::false_type) {
+    return arg.get();
+  }
+
+  template <typename T>
+  const stored_type<T>& stored_value(const T& arg, std::true_type) {
+    return storage_.push<stored_type<T>>(arg);
+  }
+
+  template <typename T> void emplace_arg(const T& arg) {
+    data_.emplace_back(internal::make_arg<Context>(arg));
+  }
+
+ public:
+  dynamic_format_arg_store() = default;
+  ~dynamic_format_arg_store() = default;
+
+  dynamic_format_arg_store(const dynamic_format_arg_store&) = delete;
+  dynamic_format_arg_store& operator=(const dynamic_format_arg_store&) = delete;
+
+  dynamic_format_arg_store(dynamic_format_arg_store&&) = default;
+  dynamic_format_arg_store& operator=(dynamic_format_arg_store&&) = default;
+
+  /**
+    \rst
+    Adds an argument into the dynamic store for later passing to a formating
+    function.
+
+    Note that custom types and string types (but not string views!) are copied
+    into the store with dynamic memory (in addition to resizing vector).
+
+    **Example**::
+
+      #include <fmt/core.h>
+      fmt::dynamic_format_arg_store<fmt::format_context> store;
+      store.push_back(42);
+      store.push_back("abc1");
+      store.push_back(1.5f);
+      std::string result = fmt::vformat("{} and {} and {}", store);
+    \endrst
+  */
+  template <typename T> void push_back(const T& arg) {
+    static_assert(
+        !std::is_base_of<internal::named_arg_base<char_type>, T>::value,
+        "Named arguments are not supported yet");
+    emplace_arg(stored_value(arg, internal::need_dyn_copy_t<T, Context>{}));
```

Comment:
i.e. replacing this with something like

```c++
if (const_check(internal::need_dyn_copy_t<T, Context>()))
  emplace_arg(storage_.push<stored_type<T>>(arg));
else
  emplace_arg(arg);
```
You'll need to move `const_check` from `fmt/format.h` to `fmt/core.h`. In the future this `if` will become `constexpr if`.

Reply:
done